Fixes #19314 - Notification for subscr. expiring soon

dLobatog opened the pull request:
Depends on https://github.com/Katello/katello/pull/6750 and https://github.com/theforeman/foreman/pull/4240/files (this https://github.com/theforeman/foreman/pull/4240/files#diff-2190144d88f9376ea36f321f8f5af34dR71 allows to set actions with more freedom than the current structure).

![screenshot from 2017-04-19 13-37-12](https://cloud.githubusercontent.com/assets/598891/25178302/9898db9e-2505-11e7-8613-a5e04a68217e.png)

ohadlevy commented on app/services/katello/ui_notifications/pulp/proxy_disk_space.rb:
```diff
@@ -0,0 +1,39 @@
+module Katello
```
file does not belongs to this pr.

ohadlevy commented on app/services/katello/ui_notifications/subscriptions/expire_soon.rb:
```diff
@@ -0,0 +1,56 @@
+                :subject => subscription,
+                :initiator => User.anonymous_admin,
+                :audience => Notification::AUDIENCE_ADMIN,
+                :message => ::UINotifications::StringParser.new(
```
if you have the expiry date of the subscription, maybe we should use it for the notification expiry date too?

ohadlevy commented on app/services/katello/ui_notifications/subscriptions/expire_soon.rb:
```diff
@@ -0,0 +1,56 @@
+        class << self
+          def deliver!
+            subs_expiring_soon.each do |subscription|
+              next if notification_already_exists?(subscription)
```
see comments on #6750 around these lines.

ohadlevy commented on app/services/katello/ui_notifications/subscriptions/expire_soon.rb:
```diff
@@ -0,0 +1,56 @@
+              ::Notification.create!(
+                :subject => subscription,
+                :initiator => User.anonymous_admin,
+                :audience => Notification::AUDIENCE_ADMIN,
```
IMHO this should be the taxonomy users/admins?

dLobatog replied:
I agree, but how do you set this? `AUDIENCE_TAXONOMY` is `subject.user_ids` - if `subject` is the subscription, it will not work. The `subject` can't be the taxonomy

ohadlevy commented on app/services/katello/ui_notifications/subscriptions/expire_soon.rb:
```diff
@@ -0,0 +1,56 @@
+
+          def subs_expiring_soon
+            result = []
+            Katello::Subscription.unscoped.all.each do |subscription|
```
is there no better query than go over all? at least I would use select instead of the next statement below.

ohadlevy commented on db/seeds.d/109-katello-notification-blueprints.rb:
```diff
@@ -0,0 +1,30 @@
+  {
+    group: _('Proxies'),
+    name: 'pulp_low_disk_space',
+    message: _("%{subject}'s disk is %{percentage} full. Since this proxy is running Pulp, it needs disk space to publish content views. Please ensure the disk does not get full."),
```
not related to this pr.

ohadlevy commented on lib/katello/tasks/proxy_disk_space_notification.rake:
```diff
@@ -0,0 +1,13 @@
+namespace :proxy_disk_space do
+  desc <<-END_DESC
+Check the proxy disk space based on Pulp API results and create notifications if
```
ditto

ohadlevy commented on lib/katello/tasks/subs_expiry_notification.rake:
```diff
@@ -0,0 +1,13 @@
+namespace :subscriptions do
+  desc <<-END_DESC
+Check the proxy disk space based on Pulp API results and create notifications if
```
wrong description titles etc.

dLobatog commented:
Updated so that:

- Notifications are expired according to the subscription expiration date
- Finding expiring_soon subs is simplified by using `.select`
- Fixed the typos in the rake task description
- Removed the Pulp low disk space notification commit (https://github.com/Katello/katello/pull/6750/files) from here.

cc @ohadlevy thanks for taking a look before

jlsherrill commented:
I'm getting an error when running the rake task, i think because Subscription is not taxable:

```
NoMethodError: undefined method `user_ids' for #<Katello::Subscription:0x00000003164bc8>
/home/vagrant/.rvm/gems/ruby-2.2.4/gems/activemodel-4.2.8/lib/active_model/attribute_methods.rb:433:in `method_missing'
/home/vagrant/git/foreman/app/models/notification.rb:43:in `subscriber_ids'
/home/vagrant/git/foreman/app/models/notification.rb:65:in `set_notification_recipients'
/home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:432:in `block in make_lambda'
```

akofink commented:
@dLobatog any update here?

ares commented:
it seems rubocop is failing, could someone with permissions set WoC? since we don't have the output in jenkins anymore, I'm retriggering tests [test]

ares commented:
this needs rebase again, ActiveJob is now in core too, but maybe it's better to wait until https://github.com/Katello/katello/pull/6750 gets in

ares commented:
LGTM, it would be great if someone from katello could take a look

jlsherrill commented on app/services/katello/ui_notifications/subscriptions/expire_soon.rb:
```diff
@@ -0,0 +1,62 @@
+              ::Notification.create!(
+                :subject => subscription,
+                :initiator => User.anonymous_admin,
+                :audience => Notification::AUDIENCE_TAXONOMY,
```
Does this require some update to foreman?  I'm getting an error that this constant isn't defined:

```
NameError: uninitialized constant Notification::AUDIENCE_TAXONOMY
	from /home/vagrant/git/foreman_hooks/lib/foreman_hooks/as_dependencies_hook.rb:4:in `load_missing_constant'
	from /home/vagrant/git/katello/app/services/katello/ui_notifications/subscriptions/expire_soon.rb:13:in `block in deliver!'
	from /home/vagrant/git/katello/app/services/katello/ui_notifications/subscriptions/expire_soon.rb:7:in `each'
	from /home/vagrant/git/katello/app/services/katello/ui_notifications/subscriptions/expire_soon.rb:7:in `deliver!'
```

ares replied:
yeah, there's no such audience, I guess we need to find administrators of a given org for which there's no good option atm, @jlsherrill can we find a user who uploaded the manifest for the organization somehow?

jlsherrill replied:
We don't track who has uploaded or refreshed a manifest.  I would suggest notifying anyone who has access to do that for the given org via the 'import_manifest' permission

ares commented:
I think it's not easy to find easily such recipients. There are no relations like this, especially if you consider filtered permissions, where the assignment might be scoped e.g. with "name ~ my". Perhaps we could only search for user who don't have such permission filtered.

ares commented:
@dLobatog I don't think you need to write custom logic, it's fine to ignore filtering conditions, but I believe it should call `user.can?(permission)` instead of `user.allowed_to?(controller_action)`, the first does not care about specific path that might change but check for permission assignment

ares commented:
@dLobatog the allowed_to example that you used only verified that user has permission to access "katello/api/v2/subscriptions/import" route based on mapping in lib/katello/permissions/. The fact you pass `organization_id` does not matter, parameters are ignored by `allowed_to` completely. In fact the result is equivalent to `user.can?(:create_subscription)` which returns true if there's any filter with the this permission assigned to the user. Even better, `can?` does not have to parse the list of routes and permissions. Using `allowed_to?` does not make sense since we are really interested in the permission, not whether user has access to specific route. Imagine there would be two routes how you can import manifest...

dLobatog commented:
@ares Actually I replied w/o looking at the code, `can?` also takes in a subject as an optiional argument, so I'll move to that. Thanks!

ares commented:
:+1: on the searching of recipients

jlsherrill commented on app/services/katello/ui_notifications/subscriptions/expire_soon.rb:
```diff
@@ -0,0 +1,62 @@
+
+          def update_expiration_date(subject)
+            expiration_date = subject.pools.
+              find { |pool| pool.expiring_soon? }.end_date.to_date
```
this can return nil if none of the pools return true on the block (i.e. no pool is found), and this causes an exception (undefined method end_date for nil class)

jlsherrill commented on app/models/katello/subscription.rb:
```diff
@@ -53,5 +55,11 @@ def virt_who?
+
+    def notification_recipients_ids
+      User.unscoped.all.find_all do |user|
+        user.can?(:import_manifest, organization_id)
```
i got an error with this, it seemed that an organization object needed to be passed in, not its id.

jlsherrill commented:
One thing to note.  I tested this in an org with 78 subscriptions most all with the same expire date.  This results in 78 notifications.  Our current notification UI does not handle lots of notifications very well and requires clicking once per notification to dismiss it.  Would it make more sense to trigger one notification per organization?

beav commented:
setting waiting on contributor due to conflicts

dLobatog commented:
[test katello]

ares commented:
[test katello]

jyejare commented:
@dLobatog, This needs rebasing.

jyejare commented:
@dLobatog , Does this PR helps to show an expiration notification before 120 days? Does the notification depend upon the place where the foreman is being accessed (e.g for me, its IST time) ?

dLobatog commented:
@jyejare Yes, you may change the 120 to any other value in `app/models/katello/pool` (DAYS_EXPIRING_SOON)  although it's not meant to be touched by users. But it can be useful when testing. You can also run `Katello::UINotifications::Subscriptions::ExpireSoon.deliver!` to ensure it runs the check - otherwise it'll only run every 12 hours.

`expiring_soon` is an old issue, it should be checking `Date.today.utc` https://github.com/Katello/katello/blob/master/app/models/katello/pool.rb#L51

dLobatog commented:
[test katello]

dLobatog commented:
https://github.com/Katello/katello/pull/7234

dLobatog commented:
I was missing - https://github.com/Katello/katello/pull/7228/files on my branch, rebasing & pushing again..

jyejare commented:
@dLobatog , @ares , I tried accessing notifications via org-admin user and via another admin, but I cant ! Looks like there is some issue with users.

dLobatog commented on app/models/katello/concerns/organization_extensions.rb:
```diff
@@ -182,6 +182,17 @@ def cancel_repo_discovery
+
+        def notification_recipients_ids
+          users = User.unscoped.all.find_all do |user|
+            user.can?(:import_manifest) && user.can?(:view_organizations, self)
```
@jyejare The problem I had was that previously I was calling user.can?(:import_manifest, self) here. This would not work, because import_manifest is a permission for Subscription, not for Organization. Hence, only admins would return 'true' for that.

tl;dr - It should be fine now for org admins

jyejare commented:
Tests Performed:

```
1. Notification should be displayed for expiring subscriptions within given days in settings.
2. Notification should only be shown for expiring subscriptions within given days in settings.
3. Notification should not be shown for subscriptions not expiring within given days in settings.
4. Notification body text should be correct.
5. Notification should be shown for org-admin users as well.
6. No duplication of notifications on multiple time delivers.
7. Pulp Notification for disk storage full with the percent of disk full.
```

QE ACK.
